drivers/usb/udc/udc_dwc3.c

```
@@ -0,0 +1,2305 @@
+	SET_BIT(trb_ptr->ctrl, USB_TRB_CTRL_HWO | USB_TRB_CTRL_LST | USB_TRB_CTRL_IOC
+			| USB_TRB_CTRL_ISP_IMI);
+
+	sys_cache_data_flush_range(trb_ptr, sizeof(*trb_ptr));
+	params.param1 = (uint32_t)trb_ptr;
+	drv->ep0_state = EP0_STATUS_PHASE;
+	/*
+	 * Control OUT transfer - Status stage happens on EP0 IN - EP1
+	 * Control IN transfer - Status stage happens on EP0 OUT - EP0
+	 */
+	dir = !drv->ep0_expect_in;
+	ret = usbd_send_ep_cmd(drv, 0U|dir, USB_DEPCMD_STARTTRANSFER, params);
+	if (ret < USB_SUCCESS) {
+		LOG_ERR("failed to execute the command at control status");
+		return ret;
+	}
+	SET_BIT(ept->ep_status, USB_EP_BUSY);
+	ept->ep_resource_index = usbd_get_ep_transfer_resource_index(drv,
+			ept->ep_index,
+			ept->ep_dir);
+
+	return ret;
+}
+
+static int32_t usbd_ep0_stall_restart(usb_dwc3_driver_t *drv, uint8_t ep_num, uint8_t dir)
+{
+	uint8_t        phy_ep;
+	usbd_ep_t       *ept    = NULL;
+	usbd_ep_params_t params = {0};
+	uint32_t       ret;
+
+	phy_ep = USB_GET_PHYSICAL_EP(ep_num, dir);
+	ept    = &drv->eps[phy_ep];
+	/* Issue the command to the hardware */
+	ret    = usbd_send_ep_cmd(drv, phy_ep, USB_DEPCMD_SETSTALL, params);
+	if (ret < 0) {
+		LOG_ERR("failed to send STALL command");
+		return ret;
+	}
+	ept->ep_status |= USB_EP_STALL;
+	/* when we issued stall on EP0 OUT after immediately
+	 * software has to schedule the setup TRB for next setup packet.
+	 */
+	usbd_prepare_setup(drv);
+	return ret;
+}
+
+void usbd_ep0_end_control_data(usb_dwc3_driver_t *drv, usbd_ep_t *ept)
+{
+	usbd_ep_params_t params = {0};
+	uint32_t     cmd;
+	uint32_t ret;
+	uint8_t phy_ep;
+
+	if (ept->ep_resource_index == 0U) {
+		return;
+	}
+	phy_ep = USB_GET_PHYSICAL_EP(ept->ep_index, ept->ep_dir);
+	cmd = USB_DEPCMD_ENDTRANSFER;
+	SET_BIT(cmd, USB_DEPCMD_CMDIOC);
+	cmd |= USB_DEPCMD_PARAM(ept->ep_resource_index);
+	ret = usbd_send_ep_cmd(drv, phy_ep, cmd, params);
+	if (ret < 0) {
+		LOG_ERR("Failed to send command at Endcontrol data");
+		return;
```

Shouldnt we propagate the error code back to the caller and make decision at the caller end